tests/integrationv2/test_external_psk.py

```diff
@@ -0,0 +1,199 @@
+    ( Ciphers.AES128_GCM_SHA256, Certificates.ECDSA_256, ['--psk', s2n_psk_parameters], ['--psk', s2n_psk_parameters_2], True ),
+    ( Ciphers.AES128_GCM_SHA256, None, ['--psk', s2n_known_value_psk], ['--psk', s2n_known_value_psk], True ),
+    ( Ciphers.AES256_GCM_SHA384, None,  ['--psk', s2n_known_value_psk_2], ['--psk', s2n_known_value_psk_2], True ),
+    ( Ciphers.AES128_GCM_SHA256, None, ['--psk', s2n_psk_parameters, '--psk', s2n_known_value_psk], ['--psk', s2n_psk_parameters_2, '--psk', s2n_known_value_psk], True )
+]
+
+OPENSSL_TEST_SUITE = [
+    ( Ciphers.AES128_GCM_SHA256, None, ['--psk', s2n_psk_parameters], openssl_psk_parameters, False ),
+    ( Ciphers.AES128_GCM_SHA256, Certificates.ECDSA_256, ['--psk', s2n_psk_parameters], openssl_psk_parameters, False ),
+    ( Ciphers.AES128_GCM_SHA256, None, ['--psk', s2n_known_value_psk], openssl_psk_parameters, True ),
+]
+
+@pytest.mark.uncollect_if(func=invalid_test_parameters)
+@pytest.mark.parametrize("test_cipher,test_certificate,test_client_params,test_server_params,test_result", S2N_TEST_SUITE)
+def test_external_psk_s2nc_with_s2nd(managed_process, test_cipher, test_certificate, test_client_params, test_server_params, test_result):
+    port = next(available_ports)
+    random_bytes = data_bytes(64)
+
+    client_options = ProviderOptions(
+        mode=S2N.ClientMode,
+        host="localhost",
+        port=port,
+        cipher=test_cipher,
+        data_to_send=random_bytes,
+        insecure=False,
+        extra_flags=test_client_params,
+        protocol=Protocols.TLS13)
+
+    if test_certificate: 
+        client_options.key=test_certificate.key
+        client_options.cert=test_certificate.cert
+        client_options.trust_store=test_certificate.cert
+
+    server_options = copy.copy(client_options)
+    server_options.data_to_send = None
+    server_options.mode = S2N.ServerMode
+    server_options.extra_flags = test_server_params
+
+    server = managed_process(S2N, server_options, timeout=5)
+    client = managed_process(S2N, client_options, timeout=5)
+
+    for results in client.get_results():
+        if test_result:
+            if test_certificate is None: 
+                assert bytes("Negotiated PSK identity: {}".format(shared_psk_identity).encode('utf-8')) in results.stdout
+            assert results.exit_code == 0
+        else:
+            assert b"Negotiated PSK identity" not in results.stdout
+            assert results.exit_code != 0
+
+    for results in server.get_results():
+        if test_result:
+            if test_certificate is None: 
+                assert bytes("Negotiated PSK identity: {}".format(shared_psk_identity).encode('utf-8')) in results.stdout
+            assert random_bytes in results.stdout
+            assert results.exit_code == 0
+        else: 
+            assert b"Negotiated PSK identity" not in results.stdout
+            assert random_bytes not in results.stdout
+            assert results.exit_code != 0
```

Comment:
You do this (with some variability on the random bytes) in every test. Do you think it's worth a helper method?

Reply:
I don't think so, it will make it difficult to read. This is the easiest to understand.

Reply:
How does factoring out common code make it more difficult to read?

Reply:
Because the common code is not very interesting it is just this line: 

```
assert bytes("Negotiated PSK identity: {}".format(shared_psk_identity).encode('utf-8')) in results.stdout
```
with addition of exit code and random_bytes when needed.

Reply:
The common code would be

Checking S2N:
```
    for results in s2n.get_results():
        if test_result:
            if test_certificate is None: 
                assert bytes("Negotiated PSK identity: {}".format(shared_psk_identity).encode('utf-8')) in results.stdout
            assert results.exit_code == 0
        else:
            assert b"Negotiated PSK identity" not in results.stdout
            assert results.exit_code != 0
```

Checking Openssl:
```
    for results in openssl.get_results():
        if result:
            assert results.exit_code == 0
        else:
            assert results.stderr is not Non
```

Checking for random bytes:
```
    for results in server.get_results():
        if test_result:
            assert random_bytes in results.stdout
        else: 
            assert random_bytes not in results.stdout
```